secretmanager/secretmanager_test.go

```diff
@@ -0,0 +1,344 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secretmanager
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1beta1"
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+	"github.com/gofrs/uuid"
+	secretspb "google.golang.org/genproto/googleapis/cloud/secrets/v1beta1"
+	grpccodes "google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
+)
+
+func testClient(tb testing.TB) (*secretmanager.Client, context.Context) {
+	tb.Helper()
+
+	ctx := context.Background()
+	client, err := secretmanager.NewClient(ctx)
+	if err != nil {
+		tb.Fatalf("failed to create client: %v", err)
+	}
+	return client, ctx
+}
+
+func testName(tb testing.TB) string {
+	tb.Helper()
+
+	u, err := uuid.NewV4()
```

Comment:
@broady just filed #1133 to avoid using UUIDs. Can we use a consistent name and avoid any leaking of resources? Maybe only create a single secret and use it for all of the tests, rather than a secret for every test?

Reply:
If we do that, tests can't run in parallel (one could be destroying while another is enabling), which is less than ideal. I can update the test helper to use a common prefix to make it easy to track down issues, but I think UUIDs are unavoidable if we want fully isolated tests and predictable tests for each sample.

Reply:
Our CI system provides a level of isolation at the cloud project level. That is, when the tests for golang-samples run, the test run has exclusive access to the project that its running in (tc.ProjectID)

So you only need to use a resource name that includes tc.ProjectID plus a pre/suffix unique to golang-samples.

Reply:
Yes, but the tests themselves (the tests in `secretmanager_test.go`) run in parallel. I can add the project ID prefix, but the secrets are also scoped to the project ID, so that seems redundant. 

For example, `TestListSecrets` runs in parallel to `TestCreateSecret`, so the results are unpredictable if we use hard-coded values. Sometimes `TestCreateSecret` would run before `TestListSecrets`, sometimes after. Similarly, some tests actually destroy secrets, so they wouldn't exist in the next run. The tests then become dependent on the project being in a certain state instead of running in isolation.

Reply:
As far as leaking resources, each function `defer` cleanup. It's theoretically possible that we'd leak if a quota issue or outage arose between the `testSecret` action and `testCleanupSecret`, but that should be incredibly rare.

Each test as written here is fully isolated and independent from all other tests and prior state of the system, by design.

Reply:
Talked with @broady about this in chat and worked through it. Worked through quota issues by switching to `SystemTest`.